File: go/fixture/go-mini/internal/handlers/authz.go
package handlers

import (
	"net/http"

	"example.com/go-mini/internal/models"
)

// requireWrite runs next only when grants carry write permission; a service
// started read-only keeps serving reads while every mutation is refused.
func requireWrite(grants models.Grants, next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if !grants.Has(models.PermWrite) {
			http.Error(w, "write permission required", http.StatusForbidden)
			return
		}
		next(w, r)
	}
}

File: py/fixture/py-mini/internal/handlers/authz.py
"""Authorization of the mutating routes."""

from http import HTTPStatus

from internal.handlers.mux import HandlerFunc, Request, Response, text
from internal.models.grants import Grants, Permission


def require_write(grants: Grants, next_handler: HandlerFunc) -> HandlerFunc:
    """Run next_handler only when grants carry write permission.

    A service started read-only keeps serving reads while every mutation is
    refused.
    """

    def handle(req: Request) -> Response:
        if not grants.has(Permission.WRITE):
            return text(HTTPStatus.FORBIDDEN, "write permission required")
        return next_handler(req)

    return handle
